tests/pykafka/test_producer.py

```diff
@@ -177,6 +180,145 @@ def test_null_payloads(self):
+        `max_request_size`
+        """
+        large_payload = b''.join([uuid4().bytes for i in range(50000)])
+
+        producer = self._get_producer(
+            auto_start=False
+        )
+        # setup producer but do not start actually sending messages
+        partition = producer._topic.partitions[0]
+        owned_broker = OwnedBroker(producer, partition.leader, auto_start=False)
+
+        delivery_report_queue = producer._cluster.handler.Queue()
+        msg = Message(
+            large_payload,
+            partition_id=0,
+            delivery_report_q=delivery_report_queue
+        )
+
+
+        owned_broker.enqueue(msg)
+
+        max_request_size = 1000
+        assert max_request_size < len(msg.value)
+        owned_broker.flush(0, max_request_size)
+        q_msg, exc = delivery_report_queue.get()
+        assert q_msg is msg
+        assert isinstance(exc, MessageSizeTooLarge)
+
+    def test_owned_broker_flush_batching_by_max_request_size(self):
+        """Test that producer batches messages into the batches no larger then
+        `max_request_size`
+        """
+        large_payload = b''.join([uuid4().bytes for i in range(5000)])
+
+        producer = self._get_producer(
+            auto_start=False
+        )
+        # setup producer but do not start actually sending messages
+        partition = producer._topic.partitions[0]
+        owned_broker = OwnedBroker(producer, partition.leader, auto_start=False)
+
+        for i in range(100):
+            msg = Message(large_payload, partition_id=0)
+            owned_broker.enqueue(msg)
+
+        batch = owned_broker.flush(0, producer._max_request_size)
+        assert len(batch) < 100
+        assert sum([len(m.value) for m in batch]) < producer._max_request_size
+
+        # iterate through the rest of the batches and test the same invariant
+        while batch:
+            batch = owned_broker.flush(0, producer._max_request_size)
+            assert len(batch) < 100
+            assert sum([len(m.value) for m in batch]) < producer._max_request_size
+
+    def test_async_produce_compression_large_message(self):
+        large_payload = b''.join([uuid4().bytes for i in range(50000)])
+        assert len(large_payload) / 1024 / 1024 < 1.0
+
+        prod = self._get_producer(
+                compression=CompressionType.SNAPPY,
+                delivery_reports=True
+                )
+        prod.produce(large_payload)
+
+        report = prod.get_delivery_report()
+        self.assertEqual(report[0].value, large_payload)
+        self.assertIsNone(report[1])
+
+        message = self.consumer.consume()
+        assert message.value == large_payload
+
+        for i in range(10):
+            prod.produce(large_payload)
+
+        reports = []
+        def ensure_all_messages_produced():
+            report = prod.get_delivery_report()
+            reports.append(report)
+            assert len(reports) == 10
```

Comment:
For me,  these assertions are a bit unintuitive - the function appends 1 report and then asserts there should be 10. I get how it works, it's mostly personal preference, so let's see what @emmett9001 thinks :)

One other downside of how `retry()` works is that the wait between the 10 calls is probably unnecessary for this particular callee, most of the time, and in that sense it slows the test down.